src/notification.c

```diff
@@ -237,6 +237,139 @@ check_value_type (const char *key,
   return FALSE;
 }
 
+static void
+markup_parser_text (GMarkupParseContext  *context,
+                    const gchar          *text,
+                    gsize                 text_len,
+                    gpointer              user_data,
+                    GError              **error)
+{
+  GString *composed = user_data;
+  gchar *str = (gchar *) text;
+
+  /* Convert new lines to spaces without adding multiple consecutive spaces */
+  while (*str)
+    {
+      gunichar c = g_utf8_get_char (str);
+      gboolean needs_space = FALSE;
+
+      while (g_unichar_isspace (c) && *str)
+        {
+          needs_space = composed->len > 0 && composed->str[composed->len - 1] != ' ';
```

Comment:
Sounds like that could do with a unit test too.

(I’m not doing a full review, so this is just a drive-by comment and I’m probably missing a lot of context. Please do ignore it if I’m off the mark here.)

Reply:
Might actually be easier to do this once after the parser is done because then you can do it in a single pass without tracking the last utf8 character.

Reply:
> You need to look at the last utf8 character which might be multiple bytes, but the only way to know how many bytes the character takes is by going through the entire string.

You can actually always find the start of the previous UTF-8 character, because the first byte in a UTF-8 codepoint encoding is distinct from the other bytes. See `g_utf8_prev_char()` or `g_utf8_find_prev_char()`. That does assume valid UTF-8 input.

Reply:
Ah, didn't know. thanks.

Reply:
https://en.wikipedia.org/wiki/UTF-8#Encoding gives a pretty good visual representation of it, which I find myself referring to quite often

Reply:
I split the out the replacement code i think it's more readable.

Reply:
I didn't add a test for invalid utf8 string since it's enforced by GVariant anyways.